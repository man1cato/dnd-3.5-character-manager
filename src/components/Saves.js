import React from 'react';

const Saves = ({saves, handleChange}) => (
    <div>
        <h3 className="row row--center">Saving Throws</h3>
        
        <div className="grid grid--6col">
            <h5 className="grid__col1">Saves</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>
            <div className="grid__col1">Fortitude</div>
            <div>{saves.fortitude.base}</div>
            <div>+</div>
            <input
                type="text"
                name="saves"
                id="fortitude"
                value={saves.fortitude.mod}
                onChange={handleChange}
            />
            <div>=</div>
            <div>{saves.fortitude.total}</div>

            <div className="grid__col1">Reflex</div>
            <div>{saves.reflex.base}</div>
            <div>+</div>
            <input
                type="text"
                name="saves"
                id="reflex"
                value={saves.reflex.mod}
                onChange={handleChange}
            />
            <div>=</div>
            <div>{saves.reflex.total}</div>

            <div className="grid__col1">Will</div>
            <div>{saves.will.base}</div>
            <div>+</div>
            <input
                type="text"
                name="saves"
                id="will"
                value={saves.will.mod}
                onChange={handleChange}
            />
            <div>=</div>
            <div>{saves.will.total}</div>
        </div>
    </div>
)

export default Saves