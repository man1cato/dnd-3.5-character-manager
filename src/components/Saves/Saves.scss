@import '../../styles/base/settings';

.Saves {
   display: grid;
   grid-template-columns: 2fr repeat(5, 1fr);
   grid-auto-flow: dense;
   grid-row-gap: $xs-size; 
   justify-items: center;
   align-items: end;
   text-align: center; 
   > input {
      max-width: 50px;
      text-align: center;
   }
}