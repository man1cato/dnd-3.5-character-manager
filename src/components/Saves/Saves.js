import React, { Fragment, useState, useEffect } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { convertInputValue } from '../../utils/utils'
import './Saves.scss'


const Saves = props => {
    const { saveBases, handleUpdate } = props
    const [saveMods, setSaveMods] = useState(props.saveMods || { fortitude: 0, reflex: 0, will: 0 })
    const [saveTotals, setSaveTotals] = useState(_.mapValues(saveMods, (val, key) => saveBases[key] + val))

    useEffect(() => {
        setSaveTotals(_.mapValues(saveBases, (val, key) => val + saveMods[key]))
        handleUpdate({ saveMods })  
    }, [saveMods])

    return (
        <div className="Saves">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(saveMods, (val, key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{saveBases[key]}</div>
                    <div>+</div>
                    <input
                        type="number"
                        data-testid={key + 'Mod'}
                        value={convertInputValue(val)}
                        onChange={(e) => {
                            setSaveMods(update(saveMods, {
                                [key]: { $set: Number(e.target.value) }
                            }))
                        }}
                    />
                    <div>=</div>
                    <div data-testid={key + 'Total'}>{saveTotals[key]}</div>
                </Fragment>
            ))}
        </div>
    )
}

export default Saves