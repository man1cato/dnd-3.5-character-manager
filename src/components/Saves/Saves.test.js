import React from 'react'
import { render, fireEvent, getNodeText } from '@testing-library/react'

import Saves from './Saves'


const handleUpdate = jest.fn()
let props

beforeEach(() => {
	props = {
		saveBases: { fortitude: 1, reflex: 4, will: 1 },
		handleUpdate
	}
})

test('should render saves with profile data', () => {
	const { container } = render(<Saves {...props} />)
	expect(container.firstChild).toMatchSnapshot()
})

test('should update total value when mod field changes', () => {
	const { getByTestId } = render(<Saves {...props} />)
	
	const value = 3
	const fortitudeModNode = getByTestId('fortitudeMod')
	fireEvent.change(fortitudeModNode, { target: { value } })
	expect(fortitudeModNode.value).toBe(`${value}`)

	const fortitudeTotalNode = getByTestId('fortitudeTotal')
	expect(getNodeText(fortitudeTotalNode)).toBe(`${props.saveBases.fortitude + value}`)
})