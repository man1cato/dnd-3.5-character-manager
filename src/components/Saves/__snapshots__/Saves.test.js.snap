// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render saves with profile data 1`] = `
<div
  class="Saves"
>
  <h5
    class="grid__col1"
  >
    Type
  </h5>
  <h5>
    Base
  </h5>
  <div />
  <h5>
    Mod
  </h5>
  <div />
  <h5>
    Total
  </h5>
  <div
    class="grid__col1"
  >
    Fortitude
  </div>
  <div>
    1
  </div>
  <div>
    +
  </div>
  <input
    data-testid="fortitudeMod"
    type="number"
    value=""
  />
  <div>
    =
  </div>
  <div
    data-testid="fortitudeTotal"
  >
    1
  </div>
  <div
    class="grid__col1"
  >
    Reflex
  </div>
  <div>
    4
  </div>
  <div>
    +
  </div>
  <input
    data-testid="reflexMod"
    type="number"
    value=""
  />
  <div>
    =
  </div>
  <div
    data-testid="reflexTotal"
  >
    4
  </div>
  <div
    class="grid__col1"
  >
    Will
  </div>
  <div>
    1
  </div>
  <div>
    +
  </div>
  <input
    data-testid="willMod"
    type="number"
    value=""
  />
  <div>
    =
  </div>
  <div
    data-testid="willTotal"
  >
    1
  </div>
</div>
`;
